File: web/src/lib/tool-calls.test.ts
import { mockHandle } from "./mock/handler";
import { mockToolCalls } from "./mock/tool-calls";
import { extraToolName, parseToolCallPage, type ToolCallText, toolCallRevision, toolCallURL } from "./tool-calls";
import type { Activity } from "./types";
import assert from "node:assert/strict";
import test from "node:test";

const event = (seq: number, kind: Activity["kind"], tool_use_id = "repeat", extra = {}): Activity => ({
  seq,
  kind,
  worker: "mainagent",
  tool: "Bash",
  tool_use_id,
  ts: "2026-09-14T00:00:00Z",
  summary: "",
  ...extra,
});

test("只响应调用和执行边界，不为普通流式文本重复查询", () => {
  const use = event(1, "tool_use");
  assert.equal(toolCallRevision([use], true), toolCallRevision([use, event(2, "text"), event(3, "thinking")], true));
  assert.notEqual(toolCallRevision([use], true), toolCallRevision([use, event(4, "tool_result")], true));
  assert.notEqual(toolCallRevision([use], true), toolCallRevision([use], false));
});

test("调用位置配对、跨页、状态筛选和稳定游标", () => {
  const events = [
    event(1, "tool_use"),
    event(2, "tool_result"),
    event(3, "tool_use"),
    event(50, "tool_result", "repeat", { is_error: true }),
    event(51, "tool_result", "orphan"),
    event(52, "tool_use", "missing"),
    event(53, "result"),
    event(54, "tool_use", "live"),
  ];
  const first = parseToolCallPage(mockToolCalls(events, [], new URLSearchParams({ limit: "2" }), "s", true));
  assert.deepEqual(
    first.items.map((c) => [c.id, c.status]),
    [
      [54, "running"],
      [52, "missing"],
    ],
  );
  assert.ok(first.has_more);
  const second = parseToolCallPage(
    mockToolCalls(events, [], new URLSearchParams({ limit: "2", cursor: first.next_cursor }), "s", true),
  );
  assert.deepEqual(
    second.items.map((c) => [c.id, c.use_id, c.result_id]),
    [
      [51, null, 51],
      [3, 3, 50],
    ],
  );
  const failed = parseToolCallPage(mockToolCalls(events, [], new URLSearchParams({ status: "failed", q: "ba" }), "s"));
  assert.deepEqual(
    failed.items.map((c) => c.id),
    [3],
  );
  assert.throws(() => mockToolCalls(events, [], new URLSearchParams({ cursor: first.next_cursor }), "other"), /游标/);
  assert.throws(
    () => mockToolCalls(events, [], new URLSearchParams({ cursor: first.next_cursor, q: "changed" }), "s"),
    /游标/,
  );
  events.push(event(55, "tool_result", "live"), event(56, "tool_use", "new"));
  const stable = parseToolCallPage(
    mockToolCalls(events, [], new URLSearchParams({ limit: "2", cursor: first.next_cursor }), "s"),
  );
  assert.deepEqual(
    stable.items.map((c) => c.id),
    [51, 3],
  );
});

test("轮次与空 ID 不按名称误配对，长中文安全分段", () => {
  const events = [
    event(1, "tool_use"),
    event(2, "result"),
    event(3, "tool_result"),
    event(4, "tool_use", ""),
    event(5, "tool_result", ""),
  ];
  const page = parseToolCallPage(mockToolCalls(events, [], new URLSearchParams(), "s"));
  assert.equal(page.items.length, 4);
  assert.equal(page.items.find((c) => c.id === 3)?.use_id, null);
  const body = "<script>不执行😀</script>".repeat(900);
  events.push(event(6, "tool_result", "long", { detail: body }));
  let rebuilt = "",
    offset = 0;
  for (;;) {
    const p = mockToolCalls(
      events,
      [],
      new URLSearchParams({ offset: String(offset), limit: "103" }),
      "s",
      false,
      6,
    ) as ToolCallText;
    rebuilt += p.text;
    if (!p.has_more) break;
    offset = p.next_offset;
  }
  assert.equal(rebuilt, body);
  assert.throws(() => mockToolCalls(events, [], new URLSearchParams(), "s", false, 99), /session/);
});

test("查询地址保留任务会话，异常响应明确报错，封装工具不伪造调用", () => {
  const u = new URL(toolCallURL("/exploration/tool-calls?task=1&session=main%3A2", {}, 3, 8000), "http://localhost");
  assert.equal(u.pathname, "/exploration/tool-calls/3");
  assert.equal(u.searchParams.get("session"), "main:2");
  assert.equal(u.searchParams.get("offset"), "8000");
  for (const value of [null, [], {}, { items: {} }]) assert.throws(() => parseToolCallPage(value), /格式错误/);
  assert.equal(extraToolName("ExecuteExtraTool", '{"tool_name":"my_custom","params":{}}'), "my_custom");
  assert.equal(extraToolName("ExecuteExtraTool", '{"tool_name":'), null);
  assert.equal(extraToolName("Bash", '{"tool_name":"pretend"}'), null);
});

test("Mock 路由覆盖两类会话，不把旧任务记录给空任务或新会话分段", async () => {
  const page = parseToolCallPage(await mockHandle("GET", "/exploration/tool-calls?task=t-acme-web&session=plan"));
  assert.ok(page.items.length > 0);
  const empty = parseToolCallPage(await mockHandle("GET", "/exploration/tool-calls?task=t-acme-api&session=plan"));
  assert.equal(empty.items.length, 0);
  const segment = parseToolCallPage(await mockHandle("GET", "/exploration/tool-calls?task=t-acme-web&session=main:99"));
  assert.equal(segment.items.length, 0);
  const chat = parseToolCallPage(await mockHandle("GET", "/conversations/1/tool-calls"));
  assert.ok(Array.isArray(chat.items));
  await assert.rejects(mockHandle("GET", "/conversations/999999/tool-calls"), /not found/);
});

File: db/tool_calls_test.go
package db

import (
	"context"
	"database/sql"
	"encoding/json"
	"errors"
	"fmt"
	"strings"
	"testing"
	"time"
)

func toolCallsFixture(t *testing.T) (*DB, ToolCallScope, func(Activity) int64) {
	t.Helper()
	d, err := Open(testDSN(t))
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(func() { d.Close() })
	c, err := d.CreateConversation("mainagent", "tool call test", nil)
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(func() { _ = d.DeleteConversation(c.ID) })
	return d, ToolCallScope{ConversationID: c.ID}, func(a Activity) int64 {
		t.Helper()
		a.Worker = "mainagent"
		id, err := d.AppendConvActivity(c.ID, a)
		if err != nil {
			t.Fatal(err)
		}
		return id
	}
}

func TestToolCallsPairingFiltersAndPaging(t *testing.T) {
	d, scope, add := toolCallsFixture(t)
	use := add(Activity{Kind: "tool_use", Tool: "Bash", ToolUseID: "same", Detail: `{"command":"first"}`})
	firstResult := add(Activity{Kind: "tool_result", Tool: "Bash", ToolUseID: "same", Detail: "first result"})
	second := add(Activity{Kind: "tool_use", Tool: "Bash", ToolUseID: "same"})
	// The next result is deliberately separated by more than an activity page.
	for range 60 {
		add(Activity{Kind: "text", Summary: "filler"})
	}
	secondResult := add(Activity{Kind: "tool_result", Tool: "Bash", ToolUseID: "same", IsError: true, Detail: "Blocked by hook"})
	orphan := add(Activity{Kind: "tool_result", Tool: "mcp__fixture__read", ToolUseID: "orphan", Detail: "orphan"})
	missing := add(Activity{Kind: "tool_use", Tool: "unknown_historical", ToolUseID: "missing"})
	add(Activity{Kind: "result"})
	running := add(Activity{Kind: "tool_use", Tool: "Bash", ToolUseID: "live"})
	query := ToolCallQuery{Limit: 2, Running: true, BuiltinNames: []string{"Bash"}}
	p, err := d.ListToolCalls(t.Context(), scope, query)
	if err != nil {
		t.Fatal(err)
	}
	if len(p.Items) != 2 || !p.HasMore || p.Items[0].ID != running || p.Items[0].Status != "running" || p.Items[1].ID != missing || p.Items[1].Status != "missing" {
		t.Fatalf("latest %+v", p)
	}
	query.Snapshot = p.SnapshotCursor
	query.Before = p.Items[1].ID
	p, err = d.ListToolCalls(t.Context(), scope, query)
	if err != nil {
		t.Fatal(err)
	}
	if len(p.Items) != 2 || p.Items[0].ID != orphan || p.Items[0].UseID != nil || p.Items[0].Type != "mcp" || p.Items[1].ID != second || *p.Items[1].ResultID != secondResult || p.Items[1].Status != "failed" {
		t.Fatalf("middle %+v", p)
	}
	query.Before = second
	p, err = d.ListToolCalls(t.Context(), scope, query)
	if err != nil {
		t.Fatal(err)
	}
	if p.HasMore || len(p.Items) != 1 || p.Items[0].ID != use || *p.Items[0].ResultID != firstResult {
		t.Fatalf("oldest %+v", p)
	}
	query = ToolCallQuery{Limit: 20, Status: "failed", Q: "bAs", Type: "builtin", BuiltinNames: []string{"Bash"}}
	p, err = d.ListToolCalls(t.Context(), scope, query)
	if err != nil || len(p.Items) != 1 || p.Items[0].ID != second {
		t.Fatalf("filters %+v %v", p, err)
	}
	query.Status = ""
	query.Q = "no match"
	p, err = d.ListToolCalls(t.Context(), scope, query)
	if err != nil || p.Items == nil || len(p.Items) != 0 || p.SnapshotCursor == 0 {
		t.Fatalf("empty %+v %v", p, err)
	}
	// A new use cannot enter a stable old page; a late result must still pair.
	snapshot := p.SnapshotCursor
	late := add(Activity{Kind: "tool_result", Tool: "Bash", ToolUseID: "live"})
	add(Activity{Kind: "tool_use", Tool: "Bash", ToolUseID: "new"})
	p, err = d.ListToolCalls(t.Context(), scope, ToolCallQuery{Limit: 20, Snapshot: snapshot})
	if err != nil {
		t.Fatal(err)
	}
	if p.Items[0].ID != running || p.Items[0].Status != "success" || *p.Items[0].ResultID != late {
		t.Fatalf("late result %+v", p)
	}
}

func TestToolCallsDetailUnicodeIsolationAndReadOnly(t *testing.T) {
	d, scope, add := toolCallsFixture(t)
	body := strings.Repeat("中文😀<script>alert(1)</script>", 1000)
	id := add(Activity{Kind: "tool_use", Tool: "Bash", ToolUseID: "long", Detail: body})
	var rebuilt strings.Builder
	for offset := 0; ; {
		p, err := d.ToolCallDetail(t.Context(), scope, id, offset, 101)
		if err != nil {
			t.Fatal(err)
		}
		rebuilt.WriteString(p.Text)
		if !p.HasMore {
			break
		}
		offset = p.NextOffset
	}
	if rebuilt.String() != body {
		t.Fatal("Unicode chunking lost data")
	}
	other, err := d.CreateConversation("mainagent", "other", nil)
	if err != nil {
		t.Fatal(err)
	}
	defer d.DeleteConversation(other.ID)
	_, err = d.ToolCallDetail(t.Context(), ToolCallScope{ConversationID: other.ID}, id, 0, 100)
	if !errors.Is(err, sql.ErrNoRows) {
		t.Fatalf("cross conversation detail: %v", err)
	}
	var before, after int
	if err = d.QueryRow(`SELECT count(*) FROM conversation_activities WHERE conversation_id=$1`, scope.ConversationID).Scan(&before); err != nil {
		t.Fatal(err)
	}
	for range 3 {
		if _, err = d.ListToolCalls(t.Context(), scope, ToolCallQuery{Limit: 20}); err != nil {
			t.Fatal(err)
		}
	}
	if err = d.QueryRow(`SELECT count(*) FROM conversation_activities WHERE conversation_id=$1`, scope.ConversationID).Scan(&after); err != nil {
		t.Fatal(err)
	}
	if before != after {
		t.Fatal("read view wrote activity")
	}
	if _, err = d.ListToolCalls(t.Context(), scope, ToolCallQuery{Limit: 51}); err == nil {
		t.Fatal("invalid limit accepted")
	}
	ctx, cancel := context.WithCancel(t.Context())
	cancel()
	if _, err = d.ListToolCalls(ctx, scope, ToolCallQuery{Limit: 20}); err == nil {
		t.Fatal("cancelled query concealed as empty")
	}
}

func TestToolCallsTaskSessionIsolation(t *testing.T) {
	d, _, _ := toolCallsFixture(t)
	exp, err := d.CreateExploration("tool session", "test")
	if err != nil {
		t.Fatal(err)
	}
	defer d.Exec(`DELETE FROM explorations WHERE id=$1`, exp)
	s := d.Exploration(exp)
	intent, err := s.AddIntent(map[string]any{"summary": "test"}, 5, nil, "planner")
	if err != nil {
		t.Fatal(err)
	}
	seg := 1
	owners := []Activity{{Worker: "mainagent"}, {Worker: "mainagent", MainSeg: &seg}, {Worker: "planner"}, {Worker: "work#1", NodeID: &intent}}
	filters := []ActivitySessionFilter{{Main: true}, {Main: true, MainSeg: &seg}, {Worker: "planner"}, {NodeID: &intent}}
	var ids []int64
	for _, a := range owners {
		a.Kind = "tool_use"
		a.ToolUseID = "reused"
		a.Tool = "Bash"
		id, err := s.AppendActivity(a)
		if err != nil {
			t.Fatal(err)
		}
		ids = append(ids, id)
	}
	for i, filter := range filters {
		p, err := d.ListToolCalls(t.Context(), s.ToolCallScope(filter), ToolCallQuery{Limit: 20})
		if err != nil {
			t.Fatal(err)
		}
		if len(p.Items) != 1 || p.Items[0].ID != ids[i] {
			t.Fatalf("session %d %+v", i, p)
		}
		_, err = d.ToolCallDetail(t.Context(), s.ToolCallScope(filter), ids[(i+1)%len(ids)], 0, 10)
		if !errors.Is(err, sql.ErrNoRows) {
			t.Fatalf("session %d leaked detail: %v", i, err)
		}
	}
}

func TestToolCallsEpochsAndCustomTypes(t *testing.T) {
	d, scope, add := toolCallsFixture(t)
	key := fmt.Sprintf("tool_call_custom_%d", scope.ConversationID)
	if err := d.SeedTool(key, "fixture", nil, nil); err != nil {
		t.Fatal(err)
	}
	if _, err := d.Exec(`UPDATE tools SET system=false,kind='command' WHERE key=$1`, key); err != nil {
		t.Fatal(err)
	}
	defer d.Exec(`DELETE FROM tools WHERE key=$1`, key)
	add(Activity{Kind: "tool_use", Tool: key, ToolUseID: "old"})
	add(Activity{Kind: "result"})
	orphan := add(Activity{Kind: "tool_result", Tool: key, ToolUseID: "old", IsError: true})
	add(Activity{Kind: "tool_use", Tool: "Bash"})
	add(Activity{Kind: "tool_result", Tool: "Bash"})
	p, err := d.ListToolCalls(t.Context(), scope, ToolCallQuery{Limit: 20})
	if err != nil {
		t.Fatal(err)
	}
	if len(p.Items) != 4 {
		t.Fatalf("missing IDs paired by name or round: %+v", p)
	}
	p, err = d.ListToolCalls(t.Context(), scope, ToolCallQuery{Limit: 20, Type: "custom", Status: "failed"})
	if err != nil || len(p.Items) != 1 || p.Items[0].ID != orphan || p.Items[0].UseID != nil {
		t.Fatalf("custom orphan %+v %v", p, err)
	}
}

func TestToolCallsLargeHistoryProjection(t *testing.T) {
	d, scope, _ := toolCallsFixture(t)
	// One SQL fixture insertion, not thousands of round trips. The list must stay
	// small even when all 2,000 result bodies are large.
	_, err := d.Exec(`INSERT INTO conversation_activities(conversation_id,worker,kind,tool,tool_use_id,detail)
	 SELECT $1,'mainagent',CASE WHEN n%2=1 THEN 'tool_use' ELSE 'tool_result' END,'Bash',((n+1)/2)::text,repeat('大',8000)
	 FROM generate_series(1,4000) n`, scope.ConversationID)
	if err != nil {
		t.Fatal(err)
	}
	started := time.Now()
	p, err := d.ListToolCalls(t.Context(), scope, ToolCallQuery{Limit: 20})
	if err != nil {
		t.Fatal(err)
	}
	body, _ := json.Marshal(p)
	if len(p.Items) != 20 || !p.HasMore || len(body) > 12000 || strings.Contains(string(body), "大") {
		t.Fatalf("unbounded list %d bytes", len(body))
	}
	t.Logf("2000 calls: %s, %d response bytes, one list SQL", time.Since(started), len(body))
	for _, count := range []int{20, 2000} {
		var plan string
		err = d.QueryRow(fmt.Sprintf(`EXPLAIN (FORMAT JSON) SELECT id FROM conversation_activities WHERE conversation_id=$1 AND kind IN ('tool_use','tool_result') ORDER BY id DESC LIMIT %d`, count), scope.ConversationID).Scan(&plan)
		if err != nil {
			t.Fatal(err)
		}
		t.Logf("query plan %d: %s", count, plan)
	}
}
